tdb/elife_upload.py

```diff
@@ -78,6 +69,41 @@ def format_measurements(self, measurements, **kwargs):
         self.disambiguate_sources(measurements)
         return measurements
 
+
+    def parse_assay_date_from_filename(self, fstem):
+        """
+        Parse assay date from the *fstem*.
+        *fstem* is expected to be formatted as `YYYYMMDD*`
+
+        If unable to parse date from *fstem*, then return masked assay date as `XXXX-XX-XX`.
+        If there are multiple valid dates, then return the latest date.
+        """
+        assay_date = "XXXX-XX-XX"
+        valid_dates = set()
+        for potential_date in re.findall(r"\d{8}", fstem):
+            # Check if the dates are valid
+            try:
+                date = datetime.datetime.strptime(potential_date, '%Y%m%d')
+            except ValueError:
+                continue
+
+            # Date is only a valid assay date if it's earlier than the current datetime!
+            if date < datetime.datetime.now():
+                valid_dates.add(date)
+
+        if len(valid_dates) == 0:
+            print(f"Failed to parse assay date from filename {fstem!r}")
+        elif len(valid_dates) == 1:
+            assay_date = datetime.datetime.strftime(valid_dates.pop(), '%Y-%m-%d')
+        else:
+            sorted_dates = list(map(lambda x: datetime.datetime.strftime(x, '%Y-%m-%d'), sorted(valid_dates)))
+            assay_date = sorted_dates[-1]
+            print(f"Found multiple potential assay dates in filename {fstem!r}: {sorted_dates}.",
+                  f"Using the last valid date as the assay date: {assay_date!r}")
```

Comment:
Does this case ever happen? Are we sure that we want to pick the latest date as the assay date when we do find multiple dates or do we want to throw an error so the user has a chance to fix the file name before the upload happens?

As a minor aesthetic note, you could also sort the dates with a list comprehension to avoid using list, map, and lambda:

```python
sorted_dates = [datetime.datetime.strftime(valid_date, '%Y-%m-%d') for valid_date in sorted(valid_dates)]
```

Reply:
> Does this case ever happen? Are we sure that we want to pick the latest date as the assay date when we do find multiple dates or do we want to throw an error so the user has a chance to fix the file name before the upload happens?

Not yet 😄 You're right that it would probably be better to throw an error here and manually fix the filename. I was thinking we would want the latest date if the file had an updated date appended, but better to be explicit about having one date in the filename.